lib/rtti.h

```diff
@@ -0,0 +1,181 @@
+// but mark upper 8 bits to be 0xFF
+template <typename T, typename = void>
+struct TypeIdResolver {
+    static constexpr TypeId resolve() noexcept {
+        // Get original 64-bit hash
+        uint64_t hash = FNV1a(TypeNameProxy<T>());
+        // Xor-fold to obtain 56-bit value
+        hash = (hash >> 56) ^ (hash & ((UINT64_C(1) << 56) - 1));
+        return (UINT64_C(0xFF) << 56) | hash;
+    }
+};
+
+// Specialized implementation that looks for static_typeId member.
+template <typename T>
+struct TypeIdResolver<T, std::void_t<decltype(T::static_typeId)>> {
+    static constexpr uint64_t resolve() noexcept {
+        constexpr TypeId id = T::static_typeId();
+        // Allow fall-back implementation if class does not want to explicitly
+        // specify a typeid
+        if constexpr (id == InvalidTypeId) return TypeIdResolver<T, int>::resolve();
+
+        return id;
+    }
+};
+
+}  // namespace detail
+
+struct Base;
+
+template <typename This, typename... Parents>
+struct TypeInfo {
+    using T = std::remove_const_t<std::remove_reference_t<This>>;
+
+    static_assert((... && std::is_base_of_v<Parents, This>),
+                  "One or more parents are not a base of this type.");
+
+    static_assert((... && std::is_base_of_v<Base, Parents>),
+                  "One or more parent hierarchies is not based on top of RTTI::Base.");
+
+    static_assert(std::is_base_of_v<Base, This>,
+                  "Invalid typeinfo requested, class is not based on top of RTTI::Base.");
+
+    static_assert(
+        std::is_same_v<T *, decltype(T::rttiEnabledMarker(nullptr))>,
+        "Invalid typeinfo requested, class does not declare typeinfo via DECLARE_TYPEINFO.");
+
+    [[nodiscard]] static constexpr TypeId id() noexcept {
+        return detail::TypeIdResolver<T>::resolve();
+    }
+
+    [[nodiscard]] static constexpr bool isA(TypeId typeId) noexcept {
+        return (id() == typeId) || (... || (Parents::TypeInfo::isA(typeId)));
+    }
+
+    template <typename T>
+    [[nodiscard]] static const void *dyn_cast(TypeId typeId, const T *ptr) noexcept {
+        if (id() == typeId) return static_cast<const This *>(ptr);
+
+        std::array<const void *, sizeof...(Parents)> ptrs = {
+            Parents::TypeInfo::dyn_cast(typeId, ptr)...};
+
+        auto it =
+            std::find_if(ptrs.begin(), ptrs.end(), [](const void *ptr) { return ptr != nullptr; });
+        return (it != ptrs.end()) ? *it : nullptr;
+    }
+};
+
+struct Base {
+    virtual ~Base() = default;
+
+    [[nodiscard]] virtual TypeId typeId() const noexcept = 0;
+    [[nodiscard]] virtual bool isA(TypeId typeId) const noexcept = 0;
+
+    template <typename T>
+    [[nodiscard]] bool is() const noexcept {
+        return isA(TypeInfo<T>::id());
+    }
+
+    template <typename T>
+    [[nodiscard]] T *to() noexcept {
+        return reinterpret_cast<T *>(const_cast<void *>(toImpl(TypeInfo<T>::id())));
```

Comment:
even simpler, would it be possible to do just `return is<T>() ? static_cast<T *>(this) : nullptr`? The compiler is fully willing to create the static cast to a child, even in presence of multiple inheritance. It will just create a nonsence pointer if the child is not really of that type.

Reply:
Well, it's not that simple sadly. We need to perform downcast here (from `Node*` to a particular `T*`). `static_cast` would work if we'd cast to the first base (of a dynamic type) as it is essentially a no-op. Things a bit more elaborated when we'd need to cast to non-first base and this is quite typical with p4c as there are lots of casts to e.g. `IDeclaration`.

Let me show you an example. Consider the following class hierarchy:
```c++
class A;
class B;
class C: public A, public B;
```

Now consider the following:
```c++
 A *a = new C;
```

For `a` downcasting to `C*` is free as it's no-op cast. In order to downcast `a` to type `B*` (or, it would be better to name this cross-cast) we need to know the exact class layout here as we need to perform so-called _this-adjustment_ (to put simple, perform something like `a + sizeof(A)`.  And this is essentially what `dynamic_cast` does: it checks the emitted typeinfo to determine offsets of bases and perform the necessary adjustment.

How can we determine the class layout and base offsets without `dynamic_cast`? Well, actually we do not need this. Compiler does this for us when virtual functions are called. Let's extend our example:
```c++
class Base {
protected:
 virtual void foo() = 0;
};
class A : public virtual Base {
protected:
  void foo() override;
};
class B : public virtual Base {
protected:
  void foo() override;
};
class C: public A, public B {
protected:
  void foo() override;
};
```

In order for everything like this to work:
```c++
A *a = new C;
B *b = new C;
C *c = new C;
a->foo(); b->foo(); c->foo();
```
the compiler would need to route the calls via so-called `this-adjustment` thunks, so inside `C::foo()` we will always get the proper `this`. And this is the trick we are using here. We are using a virtual function call to `toImpl`, so compiler would do necessary `this-adjustment` and we'll end with proper dynamic type implementation. Then we'd potentially need to perform cast to one of our bases (doing again adjustment if necessary, note that inside `dyn_cast` `T` and `This` are in direct relationship, so we ensure that `static_cast` will never produce an UB).

p4c case is slightly more elaborated. For example, we are having (this is one of the simpliest cases):
```c++
class Node : public virtual INode;
class StatOrDecl : public Node;
class IDeclaration : public virtual INode;
class Declaration : public StatOrDecl, public virtual IDeclaration;
```

If we're having code like this:
```c++
Node *n = new Declaration();
n->to<IDeclaration>();
```
then the execution flow is as follows:
 - Inside `to<T>` we determine the typeid of `IDeclaration`
 - Call virtually `toImpl`, it lands with proper `this` inside `Declaration::toImpl`; we know all immediate bases of `Declaration` via `RTTI::TypeInfo<>`.
 - From there we do upcast via `TypeInfo::dyn_cast`

Note that `TypeInfo::dyn_cast` is not virtual and all the recursion and `find` there is nicely inlined by a compiler to quite tight implementation. I tried more lazy approach (iterate over bases, return on first match) and it seems to optimize much worse provided notable slowdown.

Hope this helps understanding the flow.

I'd really like to have just plain `static_cast` like we are having in LLVM, but this is not quite possibly here :(